src/compiler/checker.rs

```diff
@@ -0,0 +1,363 @@
+            .flags()
+            .intersects(TypeFlags::StructuredOrInstantiable)
+            || target
+                .flags()
+                .intersects(TypeFlags::StructuredOrInstantiable)
+        {
+            return self.check_type_related_to(source, target, relation);
+        }
+        false
+    }
+
+    fn check_type_related_to(
+        &self,
+        source: Rc<Type>,
+        target: Rc<Type>,
+        relation: &HashMap<String, RelationComparisonResult>,
+    ) -> bool {
+        false
+    }
+
+    fn check_source_element(&self, node: &Node) {
+        self.check_source_element_worker(node)
+    }
+
+    fn check_source_element_worker(&self, node: &Node) {
+        match node {
+            Node::Statement(statement) => {
+                match statement {
+                    Statement::ExpressionStatement(expression_statement) => {
+                        return self.check_expression_statement(expression_statement);
+                    }
+                    _ => unimplemented!(),
+                };
+            }
+            _ => unimplemented!(),
+        };
+    }
+
+    fn check_source_file(&self, source_file: &SourceFile) {
+        self.check_source_file_worker(source_file)
+    }
+
+    fn check_source_file_worker(&self, node: &SourceFile) {
+        if true {
+            for_each(&node.statements, |statement, _index| {
+                self.check_source_element(statement);
+                Option::<()>::None
+            });
+        }
+    }
+
+    pub fn get_diagnostics(&self, source_file: &SourceFile) -> Vec<Rc<Diagnostic>> {
+        self.get_diagnostics_worker(source_file)
+    }
+
+    fn get_diagnostics_worker(&self, source_file: &SourceFile) -> Vec<Rc<Diagnostic>> {
+        self.check_source_file(source_file);
+
+        let semantic_diagnostics = self
+            .diagnostics()
+            .read()
+            .unwrap()
+            .get_diagnostics(&source_file.file_name);
+
+        semantic_diagnostics
+    }
+
+    fn check_arithmetic_operand_type(
+        &self,
+        operand: /*&Node*/ &Expression,
+        type_: Rc<Type>,
+        diagnostic: &DiagnosticMessage,
+    ) -> bool {
+        if !self.is_type_assignable_to(type_, self.number_or_big_int_type()) {
+            self.error_and_maybe_suggest_await(operand, diagnostic);
+            return false;
+        }
+        true
+    }
+
+    fn check_prefix_unary_expression(&self, node: &PrefixUnaryExpression) -> Rc<Type> {
+        let operand_expression = match &*node.operand {
+            Node::Expression(expression) => expression,
+            _ => panic!("Expected Expression"),
+        };
+        let operand_type = self.check_expression(operand_expression);
+        match node.operator {
+            SyntaxKind::PlusPlusToken => {
+                self.check_arithmetic_operand_type(operand_expression, operand_type.clone(), &Diagnostics::An_arithmetic_operand_must_be_of_type_any_number_bigint_or_an_enum_type);
```

Comment:
Further up the stack, this is where it's actually specifying which type error (aka "diagnostic") to emit if it thinks the argument to `++` isn't number-ish